Add dedicated email templates and controller endpoints

denarch22 opened the pull request:
## 🛠️ Changes
Introduced specific Razor views for email templates: RoleAssignment, UserRemoval, JoinInvitation, JoinRequest, and PendingRequestsReminder. Updated services to use these templates instead of the default. Extended EmailTemplatesController with endpoints to preview and test each template, and to fetch SMTP configuration for debugging.

## 📝 Associated issues
Resolve [LIB-393](https://linear.app/biodev/issue/LIB-393/monitoreo-comunitario-cambios-en-la-generacion-de-plantillas)

denarch22 committed:
Add dedicated email templates and controller endpoints

Introduced specific Razor views for email templates: RoleAssignment, UserRemoval, JoinInvitation, JoinRequest, and PendingRequestsReminder. Updated services to use these templates instead of the default. Extended EmailTemplatesController with endpoints to preview and test each template, and to fetch SMTP configuration for debugging.

cgalvist commented on src/WebApi/Controllers/Mvc/EmailTemplatesController.cs:
```diff
@@ -9,11 +17,185 @@
+    [HttpGet]
+    [Route("Config")]
+    [Produces("application/json")]
+    public IActionResult GetSmtpConfig() => Json(new
```
Esto creo que debería estar en la capa de infraestructura. Los usuarios del sistema nunca deberían poder ver las variables de entorno del servidor de correos

cgalvist commented on src/WebApi/Controllers/Mvc/EmailTemplatesController.cs:
```diff
@@ -9,11 +17,185 @@
+    /// <returns>Test result.</returns>
+    [HttpPost]
+    [Route("Test/{template}")]
+    public async Task<IActionResult> TestEmail(string template, string email, string content = null, string subject = null)
```
No sé si esta función debería estar en este ticket. Si es así, todos los comentarios deberían estar en Inglés

denarch22 replied:
Hola Cesar , ya quite esos comentarios, ya que no eran necesario, y respecto a la utilidad del archivo, es necesario para que el sistema pueda renderizar las plantillas de email (Razor) a HTML y enviarlas por correo. Sin este controlador, WebViewTools no podría encontrar y renderizar las vistas

cgalvist replied:
El controlador sí es necesario, pero el endpoint `TestEmail` que resalto aquí no se reutiliza en ninguna parte y entiendo que es para probar enviar correos y ver si las plantillas se generan correctamente, pero no es algo que debería estar en el ambiente de producción.

**EDITO:** Sí se pueden generar las plantillas sin este endpoint, haciendo los eventos respectivos para que se generen los correos electrónicos (por ejemplo, haciendo una solicitud para unirse a una iniciativa)

cgalvist commented on src/Application/Services/Initiatives/InitiativeUserService.cs:
Ya que hay plantillas independientes, los textos de `Subject` y `Content` también deberían estar en el archivo cshtml

cgalvist commented on src/Application/Services/Initiatives/InitiativeUserService.cs:
Ya que hay plantillas independientes, los textos de `Subject` y `Content` también deberían estar en el archivo cshtml

cgalvist commented on src/Application/Services/Initiatives/JoinInvitationService.cs:
Ya que hay plantillas independientes, los textos de `Subject` y `Content` también deberían estar en el archivo cshtml

cgalvist commented on src/Application/Services/Initiatives/JoinRequestService.cs:
Ya que hay plantillas independientes, los textos de `Subject` y `Content` también deberían estar en el archivo cshtml

cgalvist reviewed (changes requested):
Hola @denarch22.

Muchas gracias por tus aportes.

He dejado algunos comentarios. Quedo atento.

denarch22 committed:
Refactor email notification templates and data model

Moved email subject and content generation logic from services to Razor templates for join requests, invitations, role assignments, removals, and reminders. Extended DefaultEmailData with new properties to support template-based rendering. Removed direct string formatting from services and updated controllers and views accordingly for improved maintainability and localization.

cgalvist commented on src/Core/Domain/Utils/Email/DefaultEmailData.cs:
```diff
@@ -19,4 +19,39 @@ public class DefaultEmailData
+    /// <summary>
+    /// Initiative name (for email templates).
+    /// </summary>
+    public string InitiativeName { get; set; }
```
La clase `DefaultEmailData` solo debería tener datos por defecto de un correo electrónico. Por ejemplo, no todos los correos necesitarán el nombre de la iniciativa, o alguna de las variables de esta línea hacia abajo.
Si quieres usar esta clase como base, te recomiendo crear nuevas clases hijas de `DefaultEmailData` para cada una de las plantillas, así nos aseguramos que cada una tenga una responsabilidad única.

cgalvist reviewed (changes requested):
Hola @denarch22 .

Muchas gracias por tus aportes. Ahora el código se ve más ordenado, pero tengo algunas otras observaciones.

Quedo atento.

denarch22 committed:
Refactor email data models for template specificity

Introduced specialized email data classes for each template (JoinInvitation, JoinRequest, PendingRequestsReminder, RoleAssignment, UserRemoval) and updated service, controller, and view usages to utilize these new models. Removed template-specific properties from DefaultEmailData to improve type safety and clarity in email handling.

denarch22 committed:
Remove trailing blank lines from email data classes

Cleaned up formatting by deleting unnecessary trailing blank lines in several email data class files for consistency.

denarch22 committed:
Add BOM to email data class files

Added a UTF-8 byte order mark (BOM) to the namespace declaration in all email data class files to ensure proper encoding and compatibility with certain editors and build tools.

cgalvist commented on src/WebApi/Controllers/Mvc/EmailTemplatesController.cs:
```diff
@@ -9,11 +13,105 @@
 [ApiExplorerSettings(IgnoreApi = true)]
 public class EmailTemplatesController : Controller
 {
+    private readonly IWebViewTools webViewTools;
```
Esta interfaz no se está usando

cgalvist commented on src/WebApi/Controllers/Mvc/EmailTemplatesController.cs:
```diff
@@ -9,11 +13,105 @@
 public class EmailTemplatesController : Controller
 {
+    private readonly IWebViewTools webViewTools;
+    private readonly IEmailService emailService;
```
Esta interfaz no se está usando

cgalvist commented on src/WebApi/Controllers/Mvc/EmailTemplatesController.cs:
```diff
@@ -9,11 +13,105 @@
+    /// </summary>
+    /// <param name="webViewTools">Web view tools service.</param>
+    /// <param name="emailService">Email service.</param>
+    public EmailTemplatesController(IWebViewTools webViewTools, IEmailService emailService)
```
Ya que las interfaces no se están usando, creo que no es necesario usar este constructor

denarch22 replied:
Hola @cgalvist , ya lo ajuste pero no me di cuenta, sera que es posible agregar en el workflow una de estas validaciones para que falle en el action y evitar estos reprocesos, no se que opines

cgalvist replied:
Gracias @denarch22.
Si es posible agregar la regla, estaría de acuerdo.

denarch22 committed:
Remove unused dependencies from EmailTemplatesController

Eliminated references to IWebViewTools and IEmailService, including related constructor and fields, as they are no longer used in EmailTemplatesController.

denarch22 committed:
Enforce stricter code analysis and CI checks

Updated .editorconfig to treat several code analysis rules as errors and added new rules for unused code and code style suggestions. Enhanced the GitHub Actions workflow to run on pushes to develop and main, increased verbosity for code formatting, and improved build steps to enforce code analysis and treat warnings as errors.

cgalvist commented on .github/workflows/check-code.yml:
```diff
@@ -20,8 +24,19 @@ jobs:
-        run: dotnet build --no-incremental -warnaserror
+      - name: Build with code analysis (detects unused code)
+        run: |
+          echo "🔍 Analizando código para detectar elementos no utilizados..."
```
Porfa, coloca los textos de las acciones en Inglés

cgalvist commented on .github/workflows/check-code.yml:
```diff
@@ -20,8 +24,19 @@ jobs:
+          echo "🔍 Analizando código para detectar elementos no utilizados..."
+          dotnet build --no-incremental \
+            -warnaserror \
+            -p:TreatWarningsAsErrors=true \
```
Los parámetros `-warnaserror` y `-p:TreatWarningsAsErrors=true` son redundantes. Se puede eliminar esta línea

cgalvist commented on .github/workflows/check-code.yml:
```diff
@@ -20,8 +24,19 @@ jobs:
+            -warnaserror \
+            -p:TreatWarningsAsErrors=true \
+            -p:WarningLevel=5 \
+            -p:RunAnalyzersDuringBuild=true \
```
si este es el valor por defecto, no sería necesario este parámetro

cgalvist replied:
Esta parte quedó pendiente. Con el uso del parámetro `EnableNETAnalyzers=true` no es necesaria esta línea

cgalvist commented on .github/workflows/check-code.yml:
```diff
@@ -20,8 +24,19 @@ jobs:
+          dotnet build --no-incremental \
+            -warnaserror \
+            -p:TreatWarningsAsErrors=true \
+            -p:WarningLevel=5 \
```
si este es el valor por defecto, no sería necesario este parámetro

cgalvist commented on .github/workflows/check-code.yml:
```diff
@@ -20,8 +24,19 @@ jobs:
+            -p:WarningLevel=5 \
+            -p:RunAnalyzersDuringBuild=true \
+            -p:EnableNETAnalyzers=true \
+            -p:AnalysisLevel=8.0 \
```
Ya se está descargando en esta acción el SDK de .net 8.0, por lo que este parámetro no es necesario

cgalvist commented on .editorconfig:
```diff
@@ -167,20 +167,40 @@ dotnet_diagnostic.SA1204.severity = none
 # Remove unused private members
 # https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1822
-dotnet_diagnostic.CA1822.severity = warning
+dotnet_diagnostic.CA1822.severity = error
```
Por favor, no manejar estas reglas como errores, ya que en el ambiente de desarrollo puede generar ruido al inicio de cualquier tarea. Con el parámetro `-warnaserror` en tiempo de compilación ya se toman las advertencias como errores.

cgalvist commented on .editorconfig:
```diff
@@ -167,20 +167,40 @@ dotnet_diagnostic.SA1204.severity = none
 # Remove unused private fields
 # https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1823
-dotnet_diagnostic.CA1823.severity = warning
+dotnet_diagnostic.CA1823.severity = error
```
x2

cgalvist commented on .editorconfig:
```diff
@@ -167,20 +167,40 @@ dotnet_diagnostic.SA1204.severity = none
 # Remove unused local variables
 # https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1804
-dotnet_diagnostic.CA1804.severity = warning
+dotnet_diagnostic.CA1804.severity = error
```
x3

cgalvist commented on .editorconfig:
```diff
@@ -167,20 +167,40 @@ dotnet_diagnostic.SA1204.severity = none
 # Remove unused parameters
 # https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1801
-dotnet_diagnostic.CA1801.severity = warning
+dotnet_diagnostic.CA1801.severity = error
```
x4

cgalvist commented on .editorconfig:
```diff
@@ -167,20 +167,40 @@ dotnet_diagnostic.SA1204.severity = none
 # https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0005
-dotnet_diagnostic.IDE0005.severity = warning
\ No newline at end of file
+dotnet_diagnostic.IDE0005.severity = error
```
x5

cgalvist commented on .editorconfig:
```diff
@@ -167,20 +167,40 @@ dotnet_diagnostic.SA1204.severity = none
+dotnet_diagnostic.CA1031.severity = warning
+
+# Remove unused private methods
+# https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1811
```
No encuentro que esta regla exista

denarch22 committed:
Relax code analysis severities and update workflow

Changed several .editorconfig code analysis rules from 'error' to 'warning' to reduce build failures due to unused code. Updated GitHub Actions workflow to use English messages and simplified build analysis parameters for clarity.

cgalvist commented on .github/workflows/check-code.yml:
```diff
@@ -4,6 +4,10 @@ on:
   pull_request:
     branches:
       - develop
+  push:
```
Ahora que me pongo a pensar, esta parte tampoco es necesaria, ya que las ramas `main` y `develop` están protegidas y nadie debería poder hacer push sobre ellas

denarch22 committed:
Update CI workflow triggers and analyzer settings

Removes push triggers for develop and main branches, so the workflow now only runs on pull requests to develop. Also removes the RunAnalyzersDuringBuild property from the dotnet build step.
